apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaProducerHeadersInstrumentation.java

```diff
@@ -103,7 +103,7 @@ public static Span beforeSend(@Advice.FieldValue("apiVersions") final ApiVersion
             // https://kafka.apache.org/0110/documentation.html#messageformat
             if (apiVersions.maxUsableProduceMagic() >= RecordBatch.MAGIC_VALUE_V2 && headersSupported) {
                 try {
-                    record.headers().add(TraceContext.TRACE_PARENT_HEADER,
+                    record.headers().add(TraceContext.TRACE_PARENT_BINARY_HEADER_NAME,
                         span.getTraceContext().getOutgoingTraceParentBinaryHeader());
```

Comment:
This method returns a callback that we wrap and the span is ended only when the callback's `onCompletion` method is called, so I can't see how...

Reply:
What I would do is to have a `ThreadLocal<RecordHeader>`

Then you can do : 
```java
RecordHeader header = headerThreadLocal.get();
span.getTraceContext().fillOutgoingTraceParentBinaryHeader(header.value());
```

This minimizes allocations, as you don't have to allocate a `Header` each time, the `TraceContext` object does not get bigger in size and you also don't have to allocate a `byte[]` for the value. Additionally, there's no risk of using the value returned by `span.getTraceContext().getOutgoingTraceParentBinaryHeader()` at a wrong point in time.

The `ThreadLocal` works because after the `send` method got invoked, the header has already been serialized and sent out of process, thus it's safe to be re-used afterwards.

If you don't like

Reply:
I thought of reusing a header, but I decided not to do so as it increases dependency on internals. Both the Header implementation and relying on the headers being serialized during send (and not after - the record data is appended to some `accumulator` and handled asynchronously, I can imagine this being changed to defer headers serialization).

I must admit this is very elegant though and it looks stable since 0.11.0.0....
Let me think a bit on that